icu4c/source/common/uloc.cpp

```diff
@@ -1324,13 +1324,31 @@ _getVariant(const char* localeID,
             char prev,
             ByteSink* sink,
             const char** pEnd,
-            bool needSeparator) {
+            bool needSeparator,
+            UErrorCode& status) {
+    if (U_FAILURE(status)) return;
+    const char* start = localeID;
     bool hasVariant = false;
     if (pEnd != nullptr) { *pEnd = localeID; }
 
+    // Reasonable upper limit for variants
+    constexpr int32_t MAX_NUMBER_OF_VARIANTS = 20;
+    constexpr int32_t MAX_VARIANT_LENGTH = 8;
+    constexpr int32_t MAX_VARIANTS_LENGTH = (MAX_VARIANT_LENGTH + 1) * MAX_NUMBER_OF_VARIANTS;
+
     /* get one or more variant tags and separate them with '_' */
     if(_isIDSeparator(prev)) {
         /* get a variant string after a '-' or '_' */
+        while(!_isTerminator(*localeID)) {
+            localeID++;
+        }
+        int32_t variantsLength = localeID - start;
```

Comment:
The code would become simpler if you instead of pointer aritmetic just used a for-loop that increased an `int32_t` index variable by one upon each iteration, then you wouldn't have to store the original start of the string and calculate the length anew in each iteration.